File: snippets/screens/compose/sort_screen/2/sort_screen.kt
package com.example.compose.search

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.selection.selectable
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp

enum class SortOption(
    val key: String,
    val title: String,
    val subtitle: String,
    val icon: ImageVector,
    val hasDirection: Boolean = true
) {
    RELEVANCE("relevance", "Relevance", "Most relevant results first", Icons.Default.Star, false),
    PRICE_LOW("price_low", "Price: Low to High", "Cheapest items first", Icons.Default.KeyboardArrowUp, false),
    PRICE_HIGH("price_high", "Price: High to Low", "Most expensive items first", Icons.Default.KeyboardArrowDown, false),
    RATING("rating", "Rating", "Highest rated items first", Icons.Default.StarRate),
    NEWEST("newest", "Date Added", "Most recently added", Icons.Default.Schedule),
    POPULARITY("popularity", "Popularity", "Most popular items first", Icons.Default.TrendingUp),
    ALPHABETICAL("alphabetical", "Name", "Alphabetical order", Icons.Default.SortByAlpha),
    DISTANCE("distance", "Distance", "Nearest locations first", Icons.Default.LocationOn),
    MOST_VIEWED("most_viewed", "Most Viewed", "Highest view count first", Icons.Default.Visibility),
    MOST_DOWNLOADED("most_downloaded", "Most Downloaded", "Highest download count first", Icons.Default.Download)
}

data class SortConfiguration(
    val option: SortOption,
    val ascending: Boolean
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SortScreen(
    currentSort: SortConfiguration = SortConfiguration(SortOption.RELEVANCE, true),
    onApplySort: (SortConfiguration) -> Unit = {},
    onNavigateBack: () -> Unit = {}
) {
    var selectedSort by remember { mutableStateOf(currentSort) }

    val quickSortOptions = listOf(
        SortOption.RELEVANCE,
        SortOption.PRICE_LOW,
        SortOption.PRICE_HIGH,
        SortOption.RATING
    )

    val dateAndPopularityOptions = listOf(
        SortOption.NEWEST,
        SortOption.POPULARITY,
        SortOption.MOST_VIEWED,
        SortOption.MOST_DOWNLOADED
    )

    val otherOptions = listOf(
        SortOption.ALPHABETICAL,
        SortOption.DISTANCE
    )

    fun getSortTitle(option: SortOption, ascending: Boolean): String {
        if (!option.hasDirection) return option.title

        return when (option) {
            SortOption.RATING -> if (ascending) "Rating: Low to High" else "Rating: High to Low"
            SortOption.NEWEST -> if (ascending) "Date: Oldest First" else "Date: Newest First"
            SortOption.POPULARITY -> if (ascending) "Popularity: Low to High" else "Popularity: High to Low"
            SortOption.ALPHABETICAL -> if (ascending) "Name: A to Z" else "Name: Z to A"
            SortOption.DISTANCE -> if (ascending) "Distance: Near to Far" else "Distance: Far to Near"
            SortOption.MOST_VIEWED -> if (ascending) "Views: Low to High" else "Views: High to Low"
            SortOption.MOST_DOWNLOADED -> if (ascending) "Downloads: Low to High" else "Downloads: High to Low"
            else -> option.title
        }
    }

    fun getSortSubtitle(option: SortOption, ascending: Boolean): String {
        if (!option.hasDirection) return option.subtitle

        return when (option) {
            SortOption.RATING -> if (ascending) "Lowest rated first" else "Highest rated first"
            SortOption.NEWEST -> if (ascending) "Oldest items first" else "Newest items first"
            SortOption.POPULARITY -> if (ascending) "Least popular first" else "Most popular first"
            SortOption.ALPHABETICAL -> if (ascending) "A to Z order" else "Z to A order"
            SortOption.DISTANCE -> if (ascending) "Closest locations first" else "Farthest locations first"
            SortOption.MOST_VIEWED -> if (ascending) "Lowest views first" else "Highest views first"
            SortOption.MOST_DOWNLOADED -> if (ascending) "Lowest downloads first" else "Highest downloads first"
            else -> option.subtitle
        }
    }

    Column(
        modifier = Modifier.fillMaxSize()
    ) {
        // Top App Bar
        TopAppBar(
            title = { Text("Sort Options") },
            navigationIcon = {
                IconButton(onClick = onNavigateBack) {
                    Icon(Icons.Default.ArrowBack, contentDescription = "Back")
                }
            },
            actions = {
                TextButton(onClick = { onApplySort(selectedSort) }) {
                    Text("Apply")
                }
            }
        )

        // Content
        LazyColumn(
            modifier = Modifier.weight(1f),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            // Current Selection Summary
            if (selectedSort.option != SortOption.RELEVANCE) {
                item {
                    Card(
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.primaryContainer
                        )
                    ) {
                        Column(
                            modifier = Modifier.padding(16.dp),
                            verticalArrangement = Arrangement.spacedBy(4.dp)
                        ) {
                            Text(
                                text = "Current Sort",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )
                            Text(
                                text = getSortTitle(selectedSort.option, selectedSort.ascending),
                                style = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.Bold,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )
                            Text(
                                text = getSortSubtitle(selectedSort.option, selectedSort.ascending),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )
                        }
                    }
                }
            }

            // Quick Sort Options
            item {
                SortSection(title = "Quick Sort") {
                    quickSortOptions.forEach { option ->
                        SortOptionItem(
                            option = option,
                            isSelected = selectedSort.option == option,
                            title = getSortTitle(option, selectedSort.ascending),
                            subtitle = getSortSubtitle(option, selectedSort.ascending),
                            onClick = {
                                selectedSort = SortConfiguration(option, selectedSort.ascending)
                            }
                        )
                    }
                }
            }

            // Date & Popularity
            item {
                Divider()
            }

            item {
                SortSection(title = "Date & Popularity") {
                    dateAndPopularityOptions.forEach { option ->
                        SortOptionItem(
                            option = option,
                            isSelected = selectedSort.option == option,
                            title = getSortTitle(option, selectedSort.ascending),
                            subtitle = getSortSubtitle(option, selectedSort.ascending),
                            onClick = {
                                val defaultAscending = when (option) {
                                    SortOption.NEWEST, SortOption.POPULARITY,
                                    SortOption.MOST_VIEWED, SortOption.MOST_DOWNLOADED -> false
                                    else -> true
                                }
                                selectedSort = SortConfiguration(option, defaultAscending)
                            }
                        )
                    }
                }
            }

            // Other Options
            item {
                Divider()
            }

            item {
                SortSection(title = "Other Options") {
                    otherOptions.forEach { option ->
                        SortOptionItem(
                            option = option,
                            isSelected = selectedSort.option == option,
                            title = getSortTitle(option, selectedSort.ascending),
                            subtitle = getSortSubtitle(option, selectedSort.ascending),
                            onClick = {
                                val defaultAscending = when (option) {
                                    SortOption.ALPHABETICAL, SortOption.DISTANCE -> true
                                    else -> false
                                }
                                selectedSort = SortConfiguration(option, defaultAscending)
                            }
                        )
                    }
                }
            }

            // Sort Direction Control
            if (selectedSort.option.hasDirection) {
                item {
                    Divider()
                }

                item {
                    SortSection(title = "Sort Direction") {
                        Card {
                            Column {
                                SortDirectionItem(
                                    title = "Ascending",
                                    subtitle = getSortSubtitle(selectedSort.option, true),
                                    isSelected = selectedSort.ascending,
                                    onClick = {
                                        selectedSort = selectedSort.copy(ascending = true)
                                    }
                                )

                                Divider(modifier = Modifier.padding(start = 56.dp))

                                SortDirectionItem(
                                    title = "Descending",
                                    subtitle = getSortSubtitle(selectedSort.option, false),
                                    isSelected = !selectedSort.ascending,
                                    onClick = {
                                        selectedSort = selectedSort.copy(ascending = false)
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }

        // Bottom Actions
        Surface(
            shadowElevation = 8.dp
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                horizontalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                OutlinedButton(
                    onClick = {
                        selectedSort = SortConfiguration(SortOption.RELEVANCE, true)
                    },
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Reset")
                }

                Button(
                    onClick = { onApplySort(selectedSort) },
                    modifier = Modifier.weight(2f)
                ) {
                    Text("Apply Sort")
                }
            }
        }
    }
}

@Composable
private fun SortSection(
    title: String,
    content: @Composable ColumnScope.() -> Unit
) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.primary,
            fontWeight = FontWeight.Bold
        )
        content()
    }
}

@Composable
private fun SortOptionItem(
    option: SortOption,
    isSelected: Boolean,
    title: String,
    subtitle: String,
    onClick: () -> Unit
) {
    Card(
        onClick = onClick,
        colors = CardDefaults.cardColors(
            containerColor = if (isSelected)
                MaterialTheme.colorScheme.primaryContainer
            else
                MaterialTheme.colorScheme.surface
        )
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = option.icon,
                contentDescription = null,
                tint = if (isSelected)
                    MaterialTheme.colorScheme.onPrimaryContainer
                else
                    MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(24.dp)
            )

            Spacer(modifier = Modifier.width(16.dp))

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
                    color = if (isSelected)
                        MaterialTheme.colorScheme.onPrimaryContainer
                    else
                        MaterialTheme.colorScheme.onSurface
                )
                Text(
                    text = subtitle,
                    style = MaterialTheme.typography.bodySmall,
                    color = if (isSelected)
                        MaterialTheme.colorScheme.onPrimaryContainer
                    else
                        MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            if (isSelected) {
                Icon(
                    imageVector = Icons.Default.CheckCircle,
                    contentDescription = "Selected",
                    tint = MaterialTheme.colorScheme.onPrimaryContainer,
                    modifier = Modifier.size(20.dp)
                )
            }
        }
    }
}

@Composable
private fun SortDirectionItem(
    title: String,
    subtitle: String,
    isSelected: Boolean,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .selectable(
                selected = isSelected,
                onClick = onClick
            )
            .padding(16.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        RadioButton(
            selected = isSelected,
            onClick = onClick
        )

        Spacer(modifier = Modifier.width(16.dp))

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = title,
                style = MaterialTheme.typography.bodyLarge,
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal
            )
            Text(
                text = subtitle,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun SortScreenPreview() {
    MaterialTheme {
        SortScreen()
    }
}

File: snippets/screens/flutter/sort_screen/2/sort_screen.dart
import 'package:flutter/material.dart';

enum SortOption {
  relevance,
  priceLowToHigh,
  priceHighToLow,
  rating,
  newest,
  oldest,
  popularity,
  alphabetical,
  reverseAlphabetical,
  distance,
  mostViewed,
  mostDownloaded,
}

class SortScreen extends StatefulWidget {
  final SortOption? currentSort;

  const SortScreen({super.key, this.currentSort});

  @override
  State<SortScreen> createState() => _SortScreenState();
}

class _SortScreenState extends State<SortScreen> {
  SortOption? _selectedSort;
  bool _ascending = true;

  final Map<SortOption, Map<String, dynamic>> _sortOptions = {
    SortOption.relevance: {
      'title': 'Relevance',
      'subtitle': 'Most relevant results first',
      'icon': Icons.star,
      'hasDirection': false,
    },
    SortOption.priceLowToHigh: {
      'title': 'Price: Low to High',
      'subtitle': 'Cheapest items first',
      'icon': Icons.arrow_upward,
      'hasDirection': false,
    },
    SortOption.priceHighToLow: {
      'title': 'Price: High to Low',
      'subtitle': 'Most expensive items first',
      'icon': Icons.arrow_downward,
      'hasDirection': false,
    },
    SortOption.rating: {
      'title': 'Rating',
      'subtitle': 'Highest rated items first',
      'icon': Icons.star_rate,
      'hasDirection': true,
    },
    SortOption.newest: {
      'title': 'Date Added',
      'subtitle': 'Most recently added',
      'icon': Icons.schedule,
      'hasDirection': true,
    },
    SortOption.popularity: {
      'title': 'Popularity',
      'subtitle': 'Most popular items first',
      'icon': Icons.trending_up,
      'hasDirection': true,
    },
    SortOption.alphabetical: {
      'title': 'Name',
      'subtitle': 'Alphabetical order',
      'icon': Icons.sort_by_alpha,
      'hasDirection': true,
    },
    SortOption.distance: {
      'title': 'Distance',
      'subtitle': 'Nearest locations first',
      'icon': Icons.location_on,
      'hasDirection': true,
    },
    SortOption.mostViewed: {
      'title': 'Most Viewed',
      'subtitle': 'Highest view count first',
      'icon': Icons.visibility,
      'hasDirection': true,
    },
    SortOption.mostDownloaded: {
      'title': 'Most Downloaded',
      'subtitle': 'Highest download count first',
      'icon': Icons.download,
      'hasDirection': true,
    },
  };

  @override
  void initState() {
    super.initState();
    _selectedSort = widget.currentSort ?? SortOption.relevance;
  }

  void _applySort() {
    Navigator.pop(context, {
      'sortOption': _selectedSort,
      'ascending': _ascending,
    });
  }

  String _getSortTitle(SortOption option, bool ascending) {
    final optionData = _sortOptions[option]!;
    final title = optionData['title'] as String;

    if (!optionData['hasDirection']) return title;

    switch (option) {
      case SortOption.rating:
        return ascending ? 'Rating: Low to High' : 'Rating: High to Low';
      case SortOption.newest:
        return ascending ? 'Date: Oldest First' : 'Date: Newest First';
      case SortOption.popularity:
        return ascending ? 'Popularity: Low to High' : 'Popularity: High to Low';
      case SortOption.alphabetical:
        return ascending ? 'Name: A to Z' : 'Name: Z to A';
      case SortOption.distance:
        return ascending ? 'Distance: Near to Far' : 'Distance: Far to Near';
      case SortOption.mostViewed:
        return ascending ? 'Views: Low to High' : 'Views: High to Low';
      case SortOption.mostDownloaded:
        return ascending ? 'Downloads: Low to High' : 'Downloads: High to Low';
      default:
        return title;
    }
  }

  String _getSortSubtitle(SortOption option, bool ascending) {
    final optionData = _sortOptions[option]!;
    if (!optionData['hasDirection']) return optionData['subtitle'] as String;

    switch (option) {
      case SortOption.rating:
        return ascending ? 'Lowest rated first' : 'Highest rated first';
      case SortOption.newest:
        return ascending ? 'Oldest items first' : 'Newest items first';
      case SortOption.popularity:
        return ascending ? 'Least popular first' : 'Most popular first';
      case SortOption.alphabetical:
        return ascending ? 'A to Z order' : 'Z to A order';
      case SortOption.distance:
        return ascending ? 'Closest locations first' : 'Farthest locations first';
      case SortOption.mostViewed:
        return ascending ? 'Lowest views first' : 'Highest views first';
      case SortOption.mostDownloaded:
        return ascending ? 'Lowest downloads first' : 'Highest downloads first';
      default:
        return optionData['subtitle'] as String;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Sort Options'),
        actions: [
          TextButton(
            onPressed: _applySort,
            child: const Text('Apply'),
          ),
        ],
      ),
      body: Column(
        children: [
          // Current Selection Summary
          if (_selectedSort != null)
            Container(
              width: double.infinity,
              padding: const EdgeInsets.all(16),
              color: Theme.of(context).primaryColor.withOpacity(0.1),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    'Current Sort',
                    style: Theme.of(context).textTheme.titleSmall,
                  ),
                  const SizedBox(height: 4),
                  Text(
                    _getSortTitle(_selectedSort!, _ascending),
                    style: Theme.of(context).textTheme.titleMedium?.copyWith(
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  Text(
                    _getSortSubtitle(_selectedSort!, _ascending),
                    style: Theme.of(context).textTheme.bodySmall,
                  ),
                ],
              ),
            ),

          // Sort Options List
          Expanded(
            child: ListView(
              padding: const EdgeInsets.symmetric(vertical: 8),
              children: [
                // Quick Sort Options
                _buildSectionHeader('Quick Sort'),
                _buildSortTile(SortOption.relevance),
                _buildSortTile(SortOption.priceLowToHigh),
                _buildSortTile(SortOption.priceHighToLow),
                _buildSortTile(SortOption.rating),

                const Divider(height: 32),

                // Date & Popularity
                _buildSectionHeader('Date & Popularity'),
                _buildSortTile(SortOption.newest),
                _buildSortTile(SortOption.popularity),
                _buildSortTile(SortOption.mostViewed),
                _buildSortTile(SortOption.mostDownloaded),

                const Divider(height: 32),

                // Alphabetical & Location
                _buildSectionHeader('Other Options'),
                _buildSortTile(SortOption.alphabetical),
                _buildSortTile(SortOption.distance),

                // Sort Direction Control
                if (_selectedSort != null &&
                    _sortOptions[_selectedSort]!['hasDirection'] as bool) ...[
                  const Divider(height: 32),
                  _buildSectionHeader('Sort Direction'),
                  Card(
                    margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                    child: Column(
                      children: [
                        RadioListTile<bool>(
                          title: const Text('Ascending'),
                          subtitle: Text(_getSortSubtitle(_selectedSort!, true)),
                          value: true,
                          groupValue: _ascending,
                          onChanged: (value) {
                            setState(() => _ascending = value!);
                          },
                        ),
                        const Divider(height: 1),
                        RadioListTile<bool>(
                          title: const Text('Descending'),
                          subtitle: Text(_getSortSubtitle(_selectedSort!, false)),
                          value: false,
                          groupValue: _ascending,
                          onChanged: (value) {
                            setState(() => _ascending = value!);
                          },
                        ),
                      ],
                    ),
                  ),
                ],
              ],
            ),
          ),

          // Apply Button
          Container(
            padding: const EdgeInsets.all(16),
            decoration: BoxDecoration(
              color: Theme.of(context).scaffoldBackgroundColor,
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.1),
                  blurRadius: 4,
                  offset: const Offset(0, -2),
                ),
              ],
            ),
            child: SafeArea(
              child: Row(
                children: [
                  Expanded(
                    child: OutlinedButton(
                      onPressed: () {
                        setState(() {
                          _selectedSort = SortOption.relevance;
                          _ascending = true;
                        });
                      },
                      child: const Text('Reset'),
                    ),
                  ),
                  const SizedBox(width: 16),
                  Expanded(
                    flex: 2,
                    child: ElevatedButton(
                      onPressed: _applySort,
                      style: ElevatedButton.styleFrom(
                        padding: const EdgeInsets.all(16),
                      ),
                      child: const Text('Apply Sort'),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildSectionHeader(String title) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 16, 16, 8),
      child: Text(
        title,
        style: Theme.of(context).textTheme.titleSmall?.copyWith(
          color: Theme.of(context).primaryColor,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }

  Widget _buildSortTile(SortOption option) {
    final optionData = _sortOptions[option]!;
    final isSelected = _selectedSort == option;

    return Card(
      margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 4),
      elevation: isSelected ? 2 : 0,
      color: isSelected
          ? Theme.of(context).primaryColor.withOpacity(0.1)
          : null,
      child: RadioListTile<SortOption>(
        title: Text(
          _getSortTitle(option, _ascending),
          style: TextStyle(
            fontWeight: isSelected ? FontWeight.bold : null,
            color: isSelected ? Theme.of(context).primaryColor : null,
          ),
        ),
        subtitle: Text(
          _getSortSubtitle(option, _ascending),
          style: TextStyle(
            color: isSelected
                ? Theme.of(context).primaryColor.withOpacity(0.7)
                : null,
          ),
        ),
        secondary: Icon(
          optionData['icon'] as IconData,
          color: isSelected ? Theme.of(context).primaryColor : null,
        ),
        value: option,
        groupValue: _selectedSort,
        onChanged: (value) {
          setState(() {
            _selectedSort = value;
            // Reset to default direction when changing sort type
            if (option == SortOption.newest ||
                option == SortOption.rating ||
                option == SortOption.popularity ||
                option == SortOption.mostViewed ||
                option == SortOption.mostDownloaded) {
              _ascending = false; // Default to descending for these
            } else if (option == SortOption.alphabetical ||
                       option == SortOption.distance) {
              _ascending = true; // Default to ascending for these
            }
          });
        },
      ),
    );
  }
}